src/backend/access/heap/heapam.c

```diff
@@ -2258,7 +2258,18 @@ heap_insert(Relation relation, HeapTuple tup, CommandId cid,
 
 	END_CRIT_SECTION();
 
-	UnlockReleaseBuffer(buffer);
+	if (options & HEAP_INSERT_SPECULATIVE)
+	{
+		/*
+		 * NEON: speculative token is not stored i WAL, so we are not able to reconstruct it from hep insert WAL record.
+		 * So we have to pin this page until end of speculative insert.
+		 * Page will be unpinned by heapam_tuple_complete_speculative
+		 */
```

Comment:
```suggestion
		/*
		 * NEON: speculative token is not stored in WAL, so if the page is evicted
		 * the buffer cache, it will be lost.To prevent that, we keep the buffer pinned.
		 * It will be unpinned in heapam_tuple_complete_speculative.
		 */
```